server/controllers/gyms/getFilteredGyms.ts

```diff
@@ -0,0 +1,123 @@
+
+export default async function getFilteredGyms(req: Request, res: Response, next: NextFunction) {
+  const { name, city, typeGender, minPrice, maxPrice, availability, page, features, review } =
+    req.query;
+
+  try {
+    await filterValidation.validateAsync({ minPrice, maxPrice, page });
+
+    const where: WhereOptions<GymFilter> = {};
+
+    if (name) {
+      where.gymName = {
+        [Op.iLike]: `%${name}%`,
+      };
+    }
+    if (city) {
+      where.city = city;
+    }
+    if (typeGender) {
+      where.typeGender = typeGender;
+    }
+    if (minPrice && maxPrice) {
+      where.monthlyPrice = {
+        [Op.gte]: minPrice,
+        [Op.lte]: maxPrice,
+      };
+    } else if (minPrice) {
+      where.monthlyPrice = {
+        [Op.gte]: minPrice,
+      };
+    } else if (maxPrice) {
+      where.monthlyPrice = {
+        [Op.lte]: maxPrice,
+      };
+    }
+    if (availability) {
+      where.fulltime = availability;
+    }
+
+    if (features) {
+      const feat: string[] = (features as string).split(',');
+      where.features = {
+        [Op.contains]: feat,
+      };
+    }
+
+    let having = {};
+    if (review) {
+      having = {
+        [Op.and]: Sequelize.literal(`AVG(reviews.rate) >= ${review}`),
+      };
+    }
+
+    const gyms = await Gym.findAll({
+      attributes: [
+        'id',
+        'gymName',
+        'logo',
+        'city',
+        'typeGender',
+        'description',
+        'features',
+        'monthlyPrice',
+        'sixMonthPrice',
+        'fulltime',
+        [
+          Sequelize.literal(`(
+              SELECT 
+                CASE WHEN AVG(review.rate) IS NULL
+                      THEN 0 
+                      ELSE AVG(review.rate)
+                END
+              FROM reviews AS review     
+              WHERE
+               "review"."gymId" =  gyms.id
+              )`),
+          'review',
+        ],
+      ],
+      where,
+      having,
+      include: [
+        {
+          model: Image,
+          limit: 1,
+          required: false,
+          attributes: ['pathUrl'],
+        },
+        {
+          association: 'reviews',
+          model: Review,
+          required: false,
+          attributes: [],
+        },
+      ],
+      group: ['gyms.id'],
+      order: [[Sequelize.literal('review'), 'DESC']],
+    });
+
+    // get page from query params or default to first page
+    const currentPage: number = +page! || 1;
+    // get pager object for specified page
+    const pageSize: number = 3;
+    const pages = paginate(gyms.length, currentPage, pageSize);
+
+    // get page of items from items array
+    const pageOfGyms = gyms.slice(pages.startIndex, pages.endIndex + 1);
+
+    res.status(200).json({ gyms: pageOfGyms, pages });
+  } catch (error: any) {
+    if (error.name === 'ValidationError') {
+      next(new CustomError('عذراً خطأ في السعر أو رقم الصفحة , يجب أن يكون رقماً', 400));
```

Comment:
```suggestion
     return next(new CustomError('عذراً خطأ في السعر أو رقم الصفحة , يجب أن يكون رقماً', 400));
```
either to return or to `next(error)` in a else block